### PR TITLE
ERC4353: interface for staked NFTs (#4353)

### PR DESCRIPTION
**Pull request by aug2uag**

### CODE REVIEW

File: EIPS/eip-4353.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+## Security Considerations
+The purpose of this standard is to simply and publicly identify whether an NFT claims to have staked tokens.
+
+Staked claims will be unreliable without a locking mechanism enforced, for example, if staked tokens can only be transferred at burn. Otherwise, tokens may be deposited and/or withdrawn at any time via arbitrary methods. Also, contracts that may allow arbitrary transfers without updating the correct balance will result in potential issues. A strict rules-based approach should be taken with these edge cases in mind.
+
+A dedicated service may exist to verify the claims of a token by analyzing transactions on the explorer. In this manner, verification may be automated to ensure a token's claims are valid. The logical extension of this method may be to extend the interface and support flagging erroneous claims, all the while maintaining a simple goal of validating and verifying a staked amount exists to benefit the operator experience.
+
+## Copyright
+Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
```
Comment (MicahZoltu) [inline EIPS/eip-4353.md:235]:
```suggestion
## Copyright
Copyright and related rights waived via [CC0](../LICENSE.md).
```

File: EIPS/eip-4353.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+                {value: ethers.utils.parseEther(value)})
+            await Interface.burn(_id);
+            expect(await Interface.stakedAmount(_id)).to.equal(
+                ethers.utils.parseEther('0'));
+        });
+    });
+});
+```
+
+## Security Considerations
```
Comment (MicahZoltu) [inline EIPS/eip-4353.md:227]:
```suggestion
## Backward Compatibility
TBD

## Security Considerations
```
https://raw.githubusercontent.com/ethereum/EIPs/master/eip-template.md

File: EIPS/eip-4353.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+type: Standards Track
+category: ERC
+created: 2021-10-08
+---
+
+## Abstract
+ERC721s can be staked with tokens, but there's no means of retrieving the number of tokens staked and/or bound to an NFT. This proposal outlines a standard that may be implemented by all wallets and marketplaces easily to correctly retrieve the staked token amount of an NFT.
+
+## Motivation
+The absence of staked tokens data limits the ability of the token owner to convey to another, for any purpose including to transact, the actual amount of staked tokens as may be viewable in a wallet, marketplace, or explorer. The ability to identify and verify an exogenous value derived from the staking process may be critical to the aims of an NFT holder.
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:17]: Tried to clean up the wording a bit here:

```suggestion
Without staked token data, the actual amount of staked tokens cannot be conveyed from token owners to other users, and cannot be displayed in wallets, marketplaces, or block explorers. The ability to identify and verify an exogenous value derived from the staking process may be critical to the aims of an NFT holder.
```

File: EIPS/eip-4353.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+## Specification
+```js
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:20]:
```suggestion
```solidity
```

File: EIPS/eip-4353.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+// SPDX-License-Identifier: MIT
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:21]: We prefer CC0:

```suggestion
// SPDX-License-Identifier: CC0-1.0
```

File: EIPS/eip-4353.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+pragma solidity ^0.8.0;
+
+/**
+ * @dev Interface of the ERC165 standard, as defined in the
+ * https://eips.ethereum.org/EIPS/eip-165[EIP].
+ *
+ * Implementers can declare support of contract interfaces, which can then be
+ * queried by others ({ERC165Checker}).
+ *
+ * For an implementation, see {ERC165}.
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:32]: This comment is incorrect.

File: EIPS/eip-4353.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    
+    /**
+     * @dev Returns true if this contract implements the interface defined by
+     * `interfaceId`. See the corresponding
+     * https://eips.ethereum.org/EIPS/eip-165#how-interfaces-are-identified
+     * to learn more about how these ids are created.
+     *
+     * This function call must use less than 30 000 gas.
+     */
+    function supportsInterface(bytes4 interfaceId) external view returns (bool);
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:44]:
```suggestion
```

This is part of ERC-165, no?

File: EIPS/eip-4353.md [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+* Token Amount - the current amount of on-chain ERC20 or derived tokens bound to an NFT from one or more deposits.
+* Withdraw Mechanism - rules based approach for withdrawing staked tokens and making sure to update the balance of the staked tokens.
+
+### Staking at mint and locking tokens in NFT
+The suggested and intended implementation of this standard is to stake tokens at the time of minting an NFT, and not implementing any outbound transfer of tokens outside of `burn`. Therefore, only to stake at minting and withdraw only at burning.
+
+#### NFT displayed in wallet or marketplace
+A wallet or marketplace checks if an NFT has publicly staked tokens available for display - if so, call `stakedAmount(tokenId)` to get the current amount of tokens staked and/or bound to the NFT.
+
+The logical code looks something like this and inspired by [William Entriken](https://ethereum.stackexchange.com/a/70116/70167):
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:70]: Please remove all external links.

File: EIPS/eip-4353.md [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    function stakedAmount(uint256 tokenId) external view returns (uint256);
+    
+}
+```
+
+### Suggested flow:
+
+#### Constructor/deployment
+* Creator - the owner of an NFT with its own rules for depositing tokens at and/or after the minting of a token.
+* Token Amount - the current amount of on-chain ERC20 or derived tokens bound to an NFT from one or more deposits.
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:61]:
```suggestion
* Token Amount - the current amount of on-chain [EIP-20](./eip-20.md) or derived tokens bound to an NFT from one or more deposits.
```

File: EIPS/eip-4353.md [hunk 10]
```diff
@@ -0,0 +1,238 @@
+---
+eip: 4353
+title: Interface for Staked Tokens in NFTs
+description: This interface enables access to publicly viewable staking data of an NFT.
+author: Rex Creed (@aug2uag), Dane Scarborough <dane@nftapps.us>
+discussions-to: https://ethereum-magicians.org/t/eip-4353-viewing-staked-tokens-in-nft/7234
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-10-08
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:10]:
```suggestion
created: 2021-10-08
requires: 165
```

File: EIPS/eip-4353.md [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+     * @dev Returns uint256 amount of on-chain tokens staked to the NFT.
+     * 
+     * @dev Wallets and marketplaces would need to call this for displaying
+     *      the amount of tokens staked and/or bound to the NFT.
+     */
+    function stakedAmount(uint256 tokenId) external view returns (uint256);
+    
+}
+```
+
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:56]: It's generally a good idea to include the EIP-165 interface identifier somewhere in your EIP.

File: EIPS/eip-4353.md [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+### Staking at mint and locking tokens in NFT
+The suggested and intended implementation of this standard is to stake tokens at the time of minting an NFT, and not implementing any outbound transfer of tokens outside of `burn`. Therefore, only to stake at minting and withdraw only at burning.
+
+#### NFT displayed in wallet or marketplace
+A wallet or marketplace checks if an NFT has publicly staked tokens available for display - if so, call `stakedAmount(tokenId)` to get the current amount of tokens staked and/or bound to the NFT.
+
+The logical code looks something like this and inspired by [William Entriken](https://ethereum.stackexchange.com/a/70116/70167):
+
+```js
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:72]:
```suggestion
```solidity
```

Comment (SamWilsn) [inline EIPS/eip-4353.md:72]: This should be in the Reference Implementation section.

File: EIPS/eip-4353.md [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+
+}
+```
+
+## Rationale
+This standard is completely agnostic to how tokens are deposited or handled by the NFT. It is, therefore, the choice and responsibility of the author to encode and communicate the encoding of their tokenomics to purchasees of their token and/or to make their contracts viewable by purchasees.
+
+Although the intention of this standard is for tokens staked at mint and withdrawable only upon burn, the interface may be modified for dynamic withdrawing and depositing of tokens especially under DeFi application settings. In its current form, the contract logic may be the determining factor whether a deviation from the standard exists.
+
+## Test Case
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:149]:
```suggestion
## Test Cases
```

File: EIPS/eip-4353.md (hunk 11 above)
Comment (aug2uag) [reply EIPS/eip-4353.md:48]: is there an example you can point me to?

Comment (aug2uag) [reply EIPS/eip-4353.md:48]: updated/corrected comment and now includes `Note: The ERC-165 identifier for this interface is 0x3a3d855f.`

File: EIPS/eip-4353.md [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+A dedicated service may exist to verify the claims of a token by analyzing transactions on the explorer. In this manner, verification may be automated to ensure a token's claims are valid. The logical extension of this method may be to extend the interface and support flagging erroneous claims, all the while maintaining a simple goal of validating and verifying a staked amount exists to benefit the operator experience.
+
+## References
+
+[1] Becker S. What to know about staking — the process of locking up crypto holdings to earn rewards and interest. Business Insider. https://www.businessinsider.com/personal-finance/staking-crypto. Published 2021. Accessed June 6, 2022.
+
+[2] Entriken W. ERC721 Tokens - Hold ETH in the token?. Ethereum Stack Exchange. https://ethereum.stackexchange.com/a/70116/70167. Published 2019. Accessed October 1, 2021.
```
[the remark this answers is not in the document]
Comment (MicahZoltu) [reply EIPS/eip-4353.md]: We don't currently support anything outside of the template sections, which means no footnotes unless you embed them in an existing section (which is weird).  Recommend just inlining links to other EIPs, and removing external links.

File: EIPS/eip-4353.md [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+            await Interface.burn(_id);
+            expect(await Interface.stakedAmount(_id)).to.equal(
+                ethers.utils.parseEther('0'));
+        });
+    });
+});
+```
+
+## Backward Compatibility
+TBD
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:220]: Not sure what happened to the other comment, but `Backwards Compatibility` has to come before `Rationale` to match the order in EIP-1.

Comment (aug2uag) [reply EIPS/eip-4353.md:220]: got it

**Review by SamWilsn — APPROVED** [review]

File: EIPS/eip-4353.md [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+        super._burn(_tokenId);
+        payees[_tokenId].transfer(tokenValue[_tokenId]);
+        tokenValue[_tokenId] = 0;
+    }
+
+}
+```
+
+## Backward Compatibility
+TBD
```
Comment (SamWilsn) [inline EIPS/eip-4353.md:137]: I must have been completely asleep when I wrote my last comment. I'll open a PR to fix my own dumb mistake ^^;;

**Review by SamWilsn — CHANGES_REQUESTED** [review]

**Pull request merged**